DBZ-5119 Add `heartbeat.action.query` support for Oracle (#3513)
debezium/debezium: documentation/modules/ROOT/pages/connectors/oracle.adoc

SCN gap detection is available only if the large SCN increment occurs while the connector is running and processing near real-time events.
====

> 

Reviewer: ```suggestion
// Type: concept
// ModuleID: how-debezium-manages-offsets-in-databases-that-change-infrequently
// Title: How {prodname} manages offsets in databases that change infrequently
```
Adds modularization comments for eventual downstream use

[[low-change-frequency-offset-management]]
=== Low change frequency offset management

The {prodname} Oracle connector tracks system change numbers in the connector offsets so that when the connector is restarted, it can begin where it left off.
These offsets are part of each emitted change event; however, when the frequency of database changes are low (every few hours or days), the offsets can become stale and prevent the connector from successfully restarting if the system change number is no longer available in the transaction logs.

For connectors that use non-CDB mode to connect to Oracle, you can enable xref:oracle-property-heartbeat-interval-ms[`heartbeat.interval.ms`] to force the connector to emit a heartbeat event at regular intervals so that offsets remain synchronized.

For connectors that use CDB mode to connect to Oracle, maintaining synchronization is more complicated.

> Not only must you set xref:oracle-property-heartbeat-interval-ms[`heartbeat.interval.ms`] necessary, but it's also necessary to set xref:oracle-property-heartbeat-action-query[`heartbeat.action.query`].

Reviewer: ```suggestion
Not only must you set xref:oracle-property-heartbeat-interval-ms[`heartbeat.interval.ms`], but it's also necessary to set xref:oracle-property-heartbeat-action-query[`heartbeat.action.query`].
```

Specifying both properties is required, because in CDB mode, the connector specifically tracks changes inside the PDB only.
A supplementary mechanism is needed to trigger change events from within the pluggable database.
At regular intervals, the heartbeat action query causes the connector to insert a new table row, or update an existing row in the pluggable database.

> {prodname} detects the table changes and emits change event for them, ensuring that offsets remain synchronized, even in pluggable databases that process changes infrequently.

Reviewer: ```suggestion
{prodname} detects the table changes and emits change events for them, ensuring that offsets remain synchronized, even in pluggable databases that process changes infrequently.
```

|`0`
|Specifies, in milliseconds, how frequently the connector sends messages to a heartbeat topic. +
Use this property to determine whether the connector continues to receive change events from the source database. +
It can also be useful to set the property in situations where no change events occur in captured tables for an extended period. +
In such a case, although the connector continues to read the redo log, it emits no change event messages, so that the offset in the Kafka topic remains unchanged.
Because the connector does not flush the latest system change number (SCN) that it read from the database, the database might retain the redo log files for longer than necessary.
If the connector restarts, the extended retention period could result in the connector redundantly sending some change events. +
The default value of `0` prevents the connector from sending any heartbeat messages.

|[[oracle-property-heartbeat-action-query]]<<oracle-property-heartbeat-action-query, `+heartbeat.action.query+`>>
|No default

> ||Specifies a query that the connector executes on the source database when the connector sends a heartbeat message. +

Reviewer: ```suggestion
|Specifies a query that the connector executes on the source database when the connector sends a heartbeat message. +
```
There's an extra leading pipe character that's messing with the table formatting.

Because the connector does not flush the latest system change number (SCN) that it read from the database, the database might retain the redo log files for longer than necessary.
If the connector restarts, the extended retention period could result in the connector redundantly sending some change events. +
The default value of `0` prevents the connector from sending any heartbeat messages.

|[[oracle-property-heartbeat-action-query]]<<oracle-property-heartbeat-action-query, `+heartbeat.action.query+`>>
|No default
|Specifies a query that the connector executes on the source database when the connector sends a heartbeat message. +
 +
This is useful for resolving the situation described in xref:{link-oracle-connector}#low-change-frequency-offset-management[Low change frequency offset management], where capturing changes from a low-traffic database on the same host as a high-traffic database prevents {prodname} from synchronizing offsets. To address this situation, create a heartbeat table in the low-traffic database, and set this property to a statement that inserts records into that table, for example:  +
 +
`INSERT INTO test_heartbeat_table (text) VALUES ('test_heartbeat')` +
 +

> This allows the connector to receive changes from the low-traffic database and acknowledge their SCNs, which allows offsets to be synchronized with the broker.

Author: The wording might be a bit off from copy-n-paste from PG and changing a few words here and there.

> This allows the connector to receive changes from the low-traffic database and acknowledge their SCNs, which allows offsets to be synchronized with the broker.

Reviewer: I hear you. I was thinking of that as I made some of the changes. If you're ok with the language here, later on I can do a pass to synchronize the descriptions in the other connector docs. I figured it made sense to start somewhere and then propagate the suggested changes if they seemed ok.  I wrote/edited some of the stuff I changed now when I did my first pass on the Oracle connector doc after I had just joined the project, and I feel like what I've learned over the past year gives me a better handle on the content.

|Specifies, in milliseconds, how frequently the connector sends messages to a heartbeat topic. +
Use this property to determine whether the connector continues to receive change events from the source database. +

> It can also be useful to set the property in situations where the connector no change events occur in captured tables for an extended period. +

Reviewer: ```suggestion
It can also be useful to set the property in situations where no change events occur in captured tables for an extended period. +
```

In such a case, although the connector continues to read the redo log, it emits no change event messages, so that the offset in the Kafka topic remains unchanged.
Because the connector does not flush the latest system change number (SCN) that it read from the database, the database might retain the redo log files for longer than necessary.
If the connector restarts, the extended retention period could result in the connector redundantly sending some change events. +
The default value of `0` prevents the connector from sending any heartbeat messages.

|[[oracle-property-heartbeat-action-query]]<<oracle-property-heartbeat-action-query, `+heartbeat.action.query+`>>
|No default
||Specifies a query that the connector executes on the source database when the connector sends a heartbeat message. +

> The connector runs the query after it emits a xref:oracle-property-heartbeat-interval-ms[heartbeat message]. +

Reviewer: ```suggestion

```
Move this sentence a few lines further on, so that it follows the example.

 +
For example: +
 +
`INSERT INTO test_heartbeat_table (text) VALUES ('test_heartbeat')` +
 +

> Use this property and create a heartbeat table to resolve situations in which xref:low-change-frequency-offset-management[{prodname} fails to synchronize offsets on low-traffic databases that are on the same host as a high-traffic database].

Reviewer: ```suggestion
The connector runs the query after it emits a xref:oracle-property-heartbeat-interval-ms[heartbeat message]. 

Set this property and create a heartbeat table to receive the heartbeat messages to resolve situations in which xref:low-change-frequency-offset-management[{prodname} fails to synchronize offsets on low-traffic databases that are on the same host as a high-traffic database].
```
